### PR TITLE
feat: create new Vertex AI for Firebase plugin.

### PR DESCRIPTION
**Pull request by cynthiajoan**
## Description

Add firebase_vertexai plugin. This plugin will provide flutter developer to have access to Vertex AI service. More details in README.md of the package.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Thank you for the PR, @cynthiajoan

A few things to note:

I am aware that the timeline is quite short so I've made some suggestions. I think that this being a dev release gives us more flexibility with future changes, and it might be worth pointing out that the API could change in the future.

For instance, I'm not sure we should have a dependency on App Check as it forces the user to have it + app check plugin initialisation, particularly as we're only interested in the token. There are ways around this issue that would require some internal changes and we can think about that before making it general release.

There are a number of private methods that don't appear to be used but I'm guessing they're for future releases. In any event, they can also be removed in a future release.

I don't think the plugin needs to extend the FirebasePluginPlatform but this will cause issue with the mock tests so it can be left and removed at a later time if we wish.

File: packages/firebase_vertexai/firebase_vertexai/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Vertex AI for Firebase Flutter
+[![pub package](https://img.shields.io/pub/v/firebase_vertexai.svg)](https://pub.dev/packages/firebase_vertexai)
+
+A Flutter plugin to use the [Vertex AI](https://firebase.google.com/docs/vertex-ai/).
+
+To learn more about Vertex AI, please visit the [website](https://cloud.google.com/vertex-ai)
+
+## Getting Started
+
+To get started with Vertex AI for Firebase Flutter, please [see the documentation](https://firebase.google.com/docs/vertex-ai/get-started?platform=flutter).
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/README.md:10]: Yeah that's intended, and would be accessible when public preview is published.

File: packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_model.dart [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    required String location,
+    required FirebaseApp app,
+    FirebaseAppCheck? appCheck,
+    List<SafetySetting>? safetySettings,
+    GenerationConfig? generationConfig,
+    List<Tool>? tools,
+    Content? systemInstruction,
+    ToolConfig? toolConfig,
+  })  : _firebaseApp = app,
+        _googleAIModel = google_ai.createModelWithBaseUri(
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_model.dart:50]: Done

File: packages/firebase_vertexai/firebase_vertexai/lib/firebase_vertexai.dart [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:firebase_app_check/firebase_app_check.dart';
+import 'package:firebase_core/firebase_core.dart';
+import 'package:firebase_core_platform_interface/firebase_core_platform_interface.dart'
+    show FirebasePluginPlatform;
+import 'package:google_generative_ai/google_generative_ai.dart' as google_ai;
+// ignore: implementation_imports, tightly coupled packages
+import 'package:google_generative_ai/src/model.dart' as google_ai;
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/firebase_vertexai.dart:27]: done

File: packages/firebase_vertexai/firebase_vertexai/lib/src/firebase_vertexai.dart [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+part of firebase_vertexai;
+
+const _defaultLocation = 'us-central1';
+
+/// Default timeout duration, 30 minutes in millisecond
+const int defaultTimeout = 1800000;
+
+/// The entrypoint for [FirebaseVertexAI].
+class FirebaseVertexAI extends FirebasePluginPlatform {
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/firebase_vertexai.dart:23]: Is there any firebase common plugin related functionality will be triggered for the platform? I'm thinking about possible plugin header content etc which I'm not quite sure about

File: packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_chat.dart [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+///
+/// Records messages sent and received in [history]. The history will always
+/// record the content from the first candidate in the
+/// [GenerateContentResponse], other candidates may be available on the returned
+/// response.
+final class ChatSession {
+  /// Creates a new chat session with the provided model.
+
+  ChatSession._(
+      this._history, this._safetySettings, this._generationConfig, this._model)
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_chat.dart:27]: done

File: packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_chat.dart [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            generationConfig:
+                GenerativeModel._googleAIGenerationConfig(_generationConfig));
+  final List<Content> _history;
+  // ignore: unused_field
+  final List<SafetySetting>? _safetySettings;
+  // ignore: unused_field
+  final GenerationConfig? _generationConfig;
+  // ignore: unused_field
+  final GenerativeModel _model;
+  final google_ai.ChatSession _googleAIChatSession;
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_chat.dart:44]: done

File: packages/firebase_vertexai/firebase_vertexai/test/vertex_mock.dart [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  TestFirebaseCoreHostApi.setup(MockFirebaseAppVertexAI());
+}
+
+// FirebaseVertexAIPlatform Mock
+class MockFirebaseVertexAI extends Mock
+    with
+        // ignore: prefer_mixin, plugin_platform_interface needs to migrate to use `mixin`
+        MockPlatformInterfaceMixin
+// implements
+//     TestFirebaseStoragePlatform
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/test/vertex_mock.dart:72]: done

[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/test/vertex_mock.dart:72]: done

File: packages/firebase_vertexai/firebase_vertexai/test/vertex_mock.dart [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+{
+  MockFirebaseVertexAI() {
+    // TestFirebaseStoragePlatform();
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/test/vertex_mock.dart:75]: done

File: packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_content.dart [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+  /// The name of the function to call.
+  final String name;
+
+  /// The function parameters and values.
+  final Map<String, Object?> args;
+
+  @override
+  // TODO: Do we need the wrapper object?
+  Object toJson() => {
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/vertex_content.dart:155]: We are trying to keep it consistent with the generative ai sdk, so will be keep as it.

**Comment by cynthiajoan** [comment]
> Thank you for the PR, @cynthiajoan
> 
> A few things to note:
> 
> I am aware that the timeline is quite short so I've made some suggestions. I think that this being a dev release gives us more flexibility with future changes, and it might be worth pointing out that the API could change in the future.
> 
> For instance, I'm not sure we should have a dependency on App Check as it forces the user to have it + app check plugin initialisation, particularly as we're only interested in the token. There are ways around this issue that would require some internal changes and we can think about that before making it general release.
> 
> There are a number of private methods that don't appear to be used but I'm guessing they're for future releases. In any event, they can also be removed in a future release.
> 
> I don't think the plugin needs to extend the FirebasePluginPlatform but this will cause issue with the mock tests so it can be left and removed at a later time if we wish.

Thanks Russell. I agree with the app check part, I will point out the sdk is in preview status and apis might be changing.

File: packages/firebase_vertexai/firebase_vertexai/LICENSE [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Comment (Lyokone) [inline packages/firebase_vertexai/firebase_vertexai/LICENSE:190]: This line should be filled

File: packages/firebase_vertexai/firebase_vertexai/example/analysis_options.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  # section below, it can also be suppressed for a single line of code
+  # or a specific dart file by using the `// ignore: name_of_lint` and
+  # `// ignore_for_file: name_of_lint` syntax on the line or in the file
+  # producing the lint.
+  rules:
+    # avoid_print: false  # Uncomment to disable the `avoid_print` rule
+    # prefer_single_quotes: true  # Uncomment to enable the `prefer_single_quotes` rule
+
+# Additional information about this file can be found at
+# https://dart.dev/guides/language/analysis-options
```
Comment (Lyokone) [inline packages/firebase_vertexai/firebase_vertexai/example/analysis_options.yaml:28]: You can probably include the root analysis_options here like we did here:
packages/firebase_core/firebase_core/example/analysis_options.yaml

**Comment by Lyokone** [comment]
More on the app_check part, we can probably remove the changes made to the example App Check app, so we don't get a weird changelog for app check on the next release.

File: packages/firebase_vertexai/firebase_vertexai/LICENSE (hunk 10 above)
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/LICENSE:190]: done

File: packages/firebase_vertexai/firebase_vertexai/example/analysis_options.yaml (hunk 11 above)
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/example/analysis_options.yaml]: done

**Comment by cynthiajoan** [comment]
> More on the app_check part, we can probably remove the changes made to the example App Check app, so we don't get a weird changelog for app check on the next release. Otherwise I agree with Russell's review 👍

removed! Thanks for the review

File: packages/firebase_vertexai/firebase_vertexai/example/pubspec.yaml [hunk 12]
```diff
@@ -0,0 +1,40 @@
+name: Vertex AI for Firebase Example
```
Comment (russellwheatley) [inline packages/firebase_vertexai/firebase_vertexai/example/pubspec.yaml:1]:
```suggestion
name: vertex_ai_example
```

**Review by russellwheatley — CHANGES_REQUESTED** [review]
I think tests are failing because of the name

File: packages/firebase_vertexai/firebase_vertexai/example/pubspec.yaml (hunk 12 above)
Comment (russellwheatley) [reply packages/firebase_vertexai/firebase_vertexai/example/pubspec.yaml:1]: One more thing actually - I think @Lyokone wanted changes to be reverted to the app check example. Essentially, all the files in firebase_app_check/**/* as they aren't part of the PR. Thanks 🙏

**Review by russellwheatley — APPROVED** [review]
CI failures are unrelated 👍

**Review by a-maurice — APPROVED** [review]

File: packages/firebase_vertexai/firebase_vertexai/lib/src/firebase_vertexai.dart [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+part of firebase_vertexai;
+
+const _defaultLocation = 'us-central1';
+
+/// Default timeout duration, 30 minutes in millisecond
+const int defaultTimeout = 1800000;
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_vertexai/firebase_vertexai/lib/src/firebase_vertexai.dart:20]: It is not. However it's referenced in the test file, and I'm still learning how to access private attributes in test folder

**Pull request merged**